tests/unit/lib/intrinsic_resolver/test_intrinsic_resolver.py

```diff
@@ -0,0 +1,1443 @@
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": [delimiter, []]
+            })
+
+    @parameterized.expand([
+        ("Fn::Join should fail if the list_of_objects is not a valid list: {}".format(item), item)
+        for item in
+        [True, False, {}, 42, "t", None]
+    ])
+    def test_fn_list_of_objects_invalid_type(self, name, list_of_objects):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": ["", list_of_objects]
+            })
+
+    @parameterized.expand([
+        ("Fn::Join should require that all items in the list_of_objects resolve to string: {}".format(item),
+         item)
+        for item in
+        [True, False, {}, 42, None]
+    ])
+    def test_fn_join_items_all_str(self, name, single_obj):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": ["", ["test", single_obj, "abcd"]]
+            })
+
+
+class TestIntrinsicFnSplitResolver(TestCase):
+    def setUp(self):
+        self.resolver = IntrinsicResolver(symbol_resolver=IntrinsicsSymbolTable())
+
+    def test_basic_fn_split(self):
+        intrinsic = {
+            "Fn::Split": ["|", "a|b|c"]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, ["a", "b", "c"])
+
+    def test_nested_fn_split(self):
+        intrinsic_base_1 = {
+            "Fn::Split": [";", {"Fn::Join": [";", ["a", "b", "c"]]}]
+        }
+
+        intrinsic_base_2 = {
+            "Fn::Join": [",", intrinsic_base_1]
+        }
+        intrinsic = {
+            "Fn::Split": [",", {"Fn::Join": [",", [intrinsic_base_2, ",e", ",f,", intrinsic_base_2]]}]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, ['a', 'b', 'c', '', 'e', '', 'f', '', 'a', 'b', 'c'])
+
+    def test_delimiter_fn_split(self):
+        intrinsic_base_1 = {
+            "Fn::Join": ["", [","]]
+        }
+        intrinsic = {
+            "Fn::Split": [intrinsic_base_1, {"Fn::Join": [",", ["a", "e", "f", "d"]]}]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, ['a', 'e', 'f', 'd'])
+
+    @parameterized.expand([
+        ("Fn::Split should fail for values that are not lists: {}".format(item), item)
+        for item in
+        [True, False, "Test", {}, 42]
+    ])
+    def test_fn_split_arguments_invalid_formats(self, name, intrinsic):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Split": intrinsic
+            })
+
+    @parameterized.expand([
+        ("Fn::Split should fail if the first argument does not resolve to a string: {}".format(item), item)
```

Comment:
I don't know the answer to this question but, could I have a ref to a parameter as the first argument? Does that work in CloudFormation?